File: esami/es_grpc+jms+stomp_INVERTED/server_java/src/coda/CodaProdCons.java
package coda;

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class CodaProdCons implements ICoda {
    private final int capacity;
    private int size;
    private int testa;
    private int coda;
    private final Lock lock;
    private final Condition prodCV;
    private final Condition consCV;
    private final int[] buffer;

    public CodaProdCons(int capacity) {
        this.capacity = capacity;
        this.size = 0;
        this.testa = 0;
        this.coda = 0;
        this.buffer = new int[capacity];

        this.lock = new ReentrantLock();
        this.prodCV = lock.newCondition();
        this.consCV = lock.newCondition();
    }

    public void deposita(int item) {

        try {
            lock.lock();

            while (size == capacity) {
                try {
                    prodCV.await();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            buffer[testa] = item;
            testa = (testa + 1) % capacity;
            size += 1;

            consCV.signal();
        } finally {
            lock.unlock();
        }

    }

    public int preleva() {

        int item = -1;

        try {
            lock.lock();

            while (size == 0) {
                try {
                    consCV.await();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            item = buffer[coda];
            coda = (coda + 1) % capacity;
            size -= 1;

            prodCV.signal();
        } finally {
            lock.unlock();
        }

        return item;

    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isFull() {
        return size == capacity;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }
}

File: esami/2024-06-26_grpc+flask_REBOOT_java/manager_python/queues/codaProdCons.py
import threading as th
from queues.queue import IQueue
from queues.queueWrapper import QueueWrapper


class CodaProdCons(QueueWrapper):
    def __init__(self, queue: IQueue):
        super().__init__(queue)
        self.lock = th.Lock()
        self.prodCV = th.Condition(self.lock)
        self.consCV = th.Condition(self.lock)

    def push(self, item):
        with self.prodCV:
            while self.queue.isFull():
                self.prodCV.wait()

            self.queue.push(item)

            self.consCV.notify()

    def pop(self):
        with self.consCV:
            while self.queue.isEmpty():
                self.consCV.wait()

            item = self.queue.pop()

            self.prodCV.notify()

        return item
